libhif/hif-source.c

```diff
@@ -1657,22 +1657,53 @@ hif_source_download_package (HifSource *source,
 			     HifState *state,
 			     GError **error)
 {
+	_cleanup_ptrarray_unref_ GPtrArray *packages = g_ptr_array_new ();
+	_cleanup_free_ gchar *basename = NULL;
+
+	g_ptr_array_add (packages, pkg);
+
+	if (!hif_source_download_packages (source, packages, directory, state, error))
+		return NULL;
+
+	/* build return value */
+	basename = g_path_get_basename (hy_package_get_location (pkg));
+	return g_build_filename (directory, basename, NULL);
+}
+
+/**
+ * hif_source_download_packages:
+ * @source: a #HifSource instance.
+ * @packages: (element-type HyPackage): Array of packages, must be from this source
+ * @directory: the destination directory.
+ * @state: a #HifState.
+ * @error: a #GError or %NULL..
+ *
+ * Downloads multiple packages from a source.  The target filename
+ * will be equivalent to `g_path_get_basename (hy_package_get_location
+ * (pkg))`.
+ **/
+gboolean
+hif_source_download_packages (HifSource *source,
+			      GPtrArray *packages,
```

Comment:
HifState doesn't really give enough information, for example I couldn't do what yum does and alternate on the TTY showing [packagename] [progressbar] without being able to map from HifState -> package.  Not to mention mirrorlist failures, etc.

I think download operations are going to need a struct with detailed info.